### DISCUSSION TITLE
Anybody Have a Simple Deterministic Bevy Stage?

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#6819] [Q&A] [answered]
Hey there! I'm making a game that uses a p2p rollback networking model and I'm having an issue because of the non-deterministic system order.

There are a set of systems, the ones for handling item spawning in my game, that I really don't care what order they run in, but I need to make sure that they always run in the same order, whatever that order is.

I'm also fine running them single-threaded.

Does anybody have a simple implementation of the `Stage` trait that works like that?

I could alternatively add explicit ordering between my systems, but that's annoying because I am going to have a growing list of items,  all in different files, and trying to put an explicit ordering between them is going to be error prone as we add more items, especially when I really don't care what order they run in, just that it doesn't change between different executions of the game.

**Comment by hymm** [comment]
does `SystemStage::single_threaded` work for your use case?https://docs.rs/bevy/latest/bevy/ecs/prelude/struct.SystemStage.html#method.single_threaded

**Reply by hymm** [reply]
I suppose not since it doesn't guarantee system ordering.

**Reply by zicklag** [reply]
Yeah, that's the issue. Even though they'll run the same every time once the game starts, different runs of the game could have a different order.

**Reply by hymm** [reply]
An old version of the executor ran systems in insertion order. Not sure if it would be possible to use that or not.

**Comment by james7132** [comment, marked as answer]
Stageless should introduce a utility for setting up strict linear ordering of systems. In the meantime, https://github.com/bevyengine/bevy/pull/2381 detailed an API for tighter control over system ordering, which has been forked off into the [`bevy_system_graph`](https://crates.io/crates/bevy_system_graph) ecosystem crate.

**Reply by hymm** [reply]
Can you use bevy_system_graph to order between plugins?

**Reply by zicklag** [reply]
Very nifty @james7132, thanks for the link! That will work for me.

I'm already using some extension traits to make working with my custom rollback `Schedule` easier across plugins so I can do something similar to make the `SystemGraph` work.

**Reply by james7132** [reply]
> Can you use bevy_system_graph to order between plugins?

No, but you already have labels for those, and it's not mutually exclusive with them.
